pkg/build/generator/generator.go

```diff
@@ -127,6 +150,35 @@ func (g *BuildGenerator) createBuild(ctx kapi.Context, build *buildapi.Build) (*
 	return g.Client.GetBuild(ctx, build.Name)
 }
 
+func (g *BuildGenerator) setImageSecret(ctx kapi.Context, imageRef *kapi.ObjectReference, targetSecret *kapi.LocalObjectReference, buildNamespace string) {
+	// If the Secret is already set by user then prefer to use that instead of
+	// BuilderAccount secrets.
+	if targetSecret != nil && len(g.builderSecrets) > 0 {
+		return
+	}
+	defaultKeyring := credentialprovider.BasicDockerKeyring{}
+	for _, secret := range g.builderSecrets {
+		keyring, err := credentialprovider.MakeDockerKeyring([]kapi.Secret{secret}, &defaultKeyring)
+		if err != nil {
+			glog.V(2).Infof("Unable to make the Docker keyring for %s/%s secret: %v", secret.Name, secret.Namespace, err)
+			continue
+		}
+
+		// Get the image pull spec from the image stream reference
+		imageSpec, err := g.resolveImageStreamReference(ctx, imageRef, buildNamespace)
+		if err != nil {
+			glog.V(2).Infof("Unable to resolve the image name for %s/%s: %v", buildNamespace, imageRef, err)
+			continue
+		}
+
+		if _, found := keyring.Lookup(imageSpec); found {
+			targetSecret = &kapi.LocalObjectReference{Name: secret.Name}
```

Comment:
@deads2k `targetSecret` is pointer, agree it will be more obvious if I return it (however I will have to do the check if it is nil or not outside this function)